server/api/user_test.go

```diff
@@ -99,3 +114,165 @@ func TestCreateUser(t *testing.T) {
+			Password:      "hello1",
+			EmailVerified: true,
+			Username:      "user",
+		}
+		resp, err := th.Client.UpdateUser(&user)
+		require.Error(t, err)
+		functionaltesting.CheckUnauthorizedStatus(t, resp)
+	})
+
+	t.Run("user can't be updated by the basic user", func(t *testing.T) {
+		user := model.User{
+			Email:         "bla@gmail.com",
+			Password:      "hello1",
+			EmailVerified: true,
+			Username:      "user",
+		}
+		_, resp, err := th.UserClient.CreateUser(&user)
+		require.Error(t, err)
+		functionaltesting.CheckForbiddenStatus(t, resp)
+	})
+
+	t.Run("user can be updated by the admin", func(t *testing.T) {
+		user := model.User{
+			Email:         "bla@gmail.com",
+			Password:      "hello1",
+			EmailVerified: true,
+			Username:      "user",
+		}
+		createdUser, _, err := th.AdminClient.CreateUser(&user)
+		require.NoError(t, err)
+		createdUser.FirstName = "UpdatedName"
+		resp, err := th.AdminClient.UpdateUser(createdUser)
+		require.NoError(t, err)
+		functionaltesting.CheckOKStatus(t, resp)
+		retrievedUsers, _, err := th.AdminClient.GetUsers()
+		require.NoError(t, err)
+		require.Equal(t, (*retrievedUsers)[len(*retrievedUsers)-1].FirstName, "UpdatedName")
+	})
+
+	t.Run("user email can not be updated", func(t *testing.T) {
+		retrievedUsers, _, err := th.AdminClient.GetUsers()
+		require.NoError(t, err)
+		user := (*retrievedUsers)[len(*retrievedUsers)-1]
+		user.Email = "newMail@gmail.com"
+		resp, err := th.AdminClient.UpdateUser(&user)
+		require.Error(t, err)
+		functionaltesting.CheckForbiddenStatus(t, resp)
+	})
+}
+
+func TestDeleteUser(t *testing.T) {
+	th := functionaltesting.Setup(t)
+	defer th.TearDown()
+
+	t.Run("user can't be deleted without authentication", func(t *testing.T) {
+		userID := "123"
+		resp, err := th.Client.DeleteUser(userID)
+		require.Error(t, err)
+		functionaltesting.CheckUnauthorizedStatus(t, resp)
+	})
+
+	t.Run("user can't be updated by the basic user", func(t *testing.T) {
+		userID := "123"
+		resp, err := th.UserClient.DeleteUser(userID)
+		require.Error(t, err)
+		functionaltesting.CheckForbiddenStatus(t, resp)
+	})
+
+	t.Run("user can be deleted by the admin", func(t *testing.T) {
+		user := model.User{
+			Email:         "bla@gmail.com",
+			Password:      "hello1",
+			EmailVerified: true,
+			Username:      "user",
+		}
+		createdUser, _, err := th.AdminClient.CreateUser(&user)
+		require.NoError(t, err)
+		resp, err := th.AdminClient.DeleteUser(createdUser.ID)
+		require.NoError(t, err)
+		functionaltesting.CheckOKStatus(t, resp)
+	})
+}
+
+func TestVerifyUserEmail(t *testing.T) {
+	th := functionaltesting.Setup(t)
+	defer th.TearDown()
+
+	t.Run("email can be verified", func(t *testing.T) {
+		user := model.User{
+			Email:         "bla3@gmail.com",
+			Password:      "hello1",
+			EmailVerified: true,
+			Username:      "user3",
+		}
+		_, _, err := th.UserClient.RegisterUser(&user)
+		require.NoError(t, err)
+		token, err := th.GetLastToken()
```

Comment:
What we could do here is use `createVerifyEmailToken` to create duplicated token and verify email using it.
Another probably better way to solve this is to get token by email, we could search the `Token` table's `Extra` column with `like`

Reply:
Good idea to get a token by email 👍